src/Functions/EntropyLearnedHash.cpp

```diff
@@ -0,0 +1,291 @@
+#include <base/defines.h>
+#include <base/types.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnString.h>
+#include <Common/Exception.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context.h>
+
+/// Implementation of entropy-learned hashing: https://doi.org/10.1145/3514221.3517894
+/// If you change something in this file, please don't deviate too much from the pseudocode in the paper!
+
+/// TODOs for future work:
+/// - allow to specify an arbitrary hash function (currently always CityHash is used)
+/// - allow function chaining a la entropyLearnedHash(trainEntropyLearnedHash())
+/// - support more datatypes for data (besides String)
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_COLUMN;
+}
+
+namespace
+{
+
+using Key = String;
+using PartialKeyPositions = std::vector<size_t>;
+using Entropies = std::vector<size_t>;
+
+Key getPartialKey(std::string_view key, const PartialKeyPositions & partial_key_positions, Key & result)
+{
+    result.clear();
+    result.reserve(partial_key_positions.size());
+
+    for (auto partial_key_position : partial_key_positions)
+        if (partial_key_position < key.size())
+            result.push_back(key[partial_key_position]);
+
+    return result;
+}
+
+bool allPartialKeysAreUnique(const std::vector<Key> & data, const PartialKeyPositions & partial_key_positions)
+{
+    std::unordered_set<Key> unique_partial_keys;
+    unique_partial_keys.reserve(data.size());
+    Key partial_key;
+
+    for (const auto & key : data)
+    {
+        getPartialKey(key, partial_key_positions, partial_key);
+        if (!unique_partial_keys.insert(partial_key).second)
+            return false;
+    }
+
+    return true;
+}
+
+// NextByte returns position of byte which adds the most entropy and the new entropy
+std::pair<size_t, size_t> nextByte(const std::vector<Key> & keys, size_t max_len, PartialKeyPositions & partial_key_positions)
+{
+    size_t min_collisions = std::numeric_limits<size_t>::max();
+    size_t best_position = 0;
+
+    std::unordered_map<Key, size_t> count_table;
+    Key partial_key;
+
+    for (size_t i = 0; i < max_len; ++i)
+    {
+        count_table.clear();
+        count_table.reserve(keys.size());
+
+        partial_key_positions.push_back(i);
+        size_t collisions = 0;
+        for (const auto & key : keys)
+        {
+            getPartialKey(key, partial_key_positions, partial_key);
+            collisions += count_table[partial_key]++;
```

Comment:
also I would consider comparing hashes instead of strings itself. we can update hash in `O(1)` and compare hashes in `O(1)`.